File: examples/security-reviewer/python/validate.py
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

Severity = Literal["critical", "high", "medium"]

EXAMPLE_ROOT = Path(__file__).resolve().parents[1]


@dataclass(frozen=True)
class SecurityFinding:
    rule: str
    severity: Severity
    title: str
    path: str
    detail: str


def resolve_diff_path(override: str | None = None) -> Path:
    if override and override.strip():
        candidate = Path(override.strip())
        return candidate if candidate.is_absolute() else Path.cwd() / candidate
    return EXAMPLE_ROOT / "fixtures" / "pr-auth-bypass.diff"


def list_changed_files(diff: str) -> list[str]:
    files: list[str] = []
    seen: set[str] = set()
    for line in diff.splitlines():
        match = re.match(r"^\+\+\+ b/(.+)$", line)
        if match and match.group(1) != "/dev/null":
            path = match.group(1)
            if path not in seen:
                seen.add(path)
                files.append(path)
    return files


def added_lines_for_file(diff: str, file_path: str) -> list[str]:
    lines: list[str] = []
    current_file: str | None = None
    in_hunk = False
    for line in diff.splitlines():
        file_match = re.match(r"^\+\+\+ b/(.+)$", line)
        if file_match:
            current_file = file_match.group(1)
            in_hunk = False
            continue
        if line.startswith("@@"):
            in_hunk = current_file == file_path
            continue
        if not in_hunk or current_file != file_path:
            continue
        if line.startswith("+") and not line.startswith("+++"):
            lines.append(line[1:])
    return lines


def validate_security_reviewer_diff(diff: str) -> list[SecurityFinding]:
    findings: list[SecurityFinding] = []

    session_text = "\n".join(added_lines_for_file(diff, "src/auth/session.ts"))
    if re.search(
        r"if\s*\(\s*!session\s*\)[\s\S]*return true", session_text
    ) or re.search(r"isAdmin\([\s\S]*null[\s\S]*return true", session_text, re.S):
        findings.append(
            SecurityFinding(
                rule="auth-bypass",
                severity="critical",
                title="Admin check treats missing session as admin",
                path="src/auth/session.ts",
                detail="isAdmin returns true when session is null. Missing auth must deny access.",
            )
        )

    refund_text = "\n".join(added_lines_for_file(diff, "src/payments/refund.ts"))
    if re.search(r"sk_live_|api[_-]?key\s*=\s*['\"][^'\"]+['\"]", refund_text, re.I):
        findings.append(
            SecurityFinding(
                rule="hardcoded-secret",
                severity="critical",
                title="Live API key committed to source",
                path="src/payments/refund.ts",
                detail="Gateway credential appears in source. Load from a secret manager.",
            )
        )

    if re.search(r"`[^`]*\$\{[^}]+\}[^`]*`", refund_text) and re.search(
        r"SELECT|INSERT|UPDATE", refund_text, re.I
    ):
        findings.append(
            SecurityFinding(
                rule="sql-injection",
                severity="high",
                title="User input concatenated into SQL",
                path="src/payments/refund.ts",
                detail="Build SQL with parameterized queries instead of template literals.",
            )
        )

    admin_text = "\n".join(added_lines_for_file(diff, "src/admin/tools.ts"))
    if re.search(r"\beval\s*\(", admin_text):
        findings.append(
            SecurityFinding(
                rule="dangerous-eval",
                severity="critical",
                title="Agent tool executes arbitrary shell via eval",
                path="src/admin/tools.ts",
                detail="Admin tool uses eval on user-supplied command text. Remove or gate behind human approval.",
            )
        )

    return findings


def format_findings(findings: list[SecurityFinding]) -> str:
    if not findings:
        return "No Security Reviewer rule violations detected in the fixture diff."
    return "\n\n".join(
        f"[{finding.severity}] {finding.title}\n"
        f"  rule: {finding.rule}\n"
        f"  path: {finding.path}\n"
        f"  {finding.detail}"
        for finding in findings
    )

File: examples/security-reviewer/ts/src/validate.ts
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

export type SecurityFinding = {
  rule: string;
  severity: "critical" | "high" | "medium";
  title: string;
  path: string;
  detail: string;
};

const exampleRoot = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  "../..",
);

export function resolveDiffPath(override?: string): string {
  const trimmed = override?.trim();
  if (trimmed) {
    return path.isAbsolute(trimmed)
      ? trimmed
      : path.resolve(process.cwd(), trimmed);
  }

  return path.join(exampleRoot, "fixtures/pr-auth-bypass.diff");
}

export function listChangedFiles(diff: string): string[] {
  const files = new Set<string>();

  for (const line of diff.split("\n")) {
    const match = line.match(/^\+\+\+ b\/(.+)$/);
    if (match?.[1] && match[1] !== "/dev/null") {
      files.add(match[1]);
    }
  }

  return [...files];
}

export function addedLinesForFile(diff: string, filePath: string): string[] {
  const lines: string[] = [];
  let currentFile: string | null = null;
  let inHunk = false;

  for (const line of diff.split("\n")) {
    const fileMatch = line.match(/^\+\+\+ b\/(.+)$/);
    if (fileMatch?.[1]) {
      currentFile = fileMatch[1];
      inHunk = false;
      continue;
    }

    if (line.startsWith("@@")) {
      inHunk = currentFile === filePath;
      continue;
    }

    if (!inHunk || currentFile !== filePath) {
      continue;
    }

    if (line.startsWith("+") && !line.startsWith("+++")) {
      lines.push(line.slice(1));
    }
  }

  return lines;
}

export function validateSecurityReviewerDiff(diff: string): SecurityFinding[] {
  const findings: SecurityFinding[] = [];

  const sessionAdds = addedLinesForFile(diff, "src/auth/session.ts").join("\n");
  if (
    /isAdmin\([\s\S]*null[\s\S]*return true/s.test(sessionAdds) ||
    /if\s*\(\s*!session\s*\)[\s\S]*return true/s.test(sessionAdds)
  ) {
    findings.push({
      rule: "auth-bypass",
      severity: "critical",
      title: "Admin check treats missing session as admin",
      path: "src/auth/session.ts",
      detail:
        "isAdmin returns true when session is null. Missing auth must deny access.",
    });
  }

  const refundAdds = addedLinesForFile(diff, "src/payments/refund.ts").join("\n");
  if (/sk_live_|api[_-]?key\s*=\s*["'][^"']+["']/i.test(refundAdds)) {
    findings.push({
      rule: "hardcoded-secret",
      severity: "critical",
      title: "Live API key committed to source",
      path: "src/payments/refund.ts",
      detail: "Gateway credential appears in source. Load from a secret manager.",
    });
  }

  if (/`[^`]*\$\{[^}]+\}[^`]*`/.test(refundAdds) && /SELECT|INSERT|UPDATE/i.test(refundAdds)) {
    findings.push({
      rule: "sql-injection",
      severity: "high",
      title: "User input concatenated into SQL",
      path: "src/payments/refund.ts",
      detail: "Build SQL with parameterized queries instead of template literals.",
    });
  }

  const adminAdds = addedLinesForFile(diff, "src/admin/tools.ts").join("\n");
  if (/\beval\s*\(/.test(adminAdds)) {
    findings.push({
      rule: "dangerous-eval",
      severity: "critical",
      title: "Agent tool executes arbitrary shell via eval",
      path: "src/admin/tools.ts",
      detail:
        "Admin tool uses eval on user-supplied command text. Remove or gate behind human approval.",
    });
  }

  return findings;
}

export async function loadDefaultDiff(): Promise<string> {
  return fs.readFile(resolveDiffPath(), "utf8");
}

export function formatFindings(findings: SecurityFinding[]): string {
  if (findings.length === 0) {
    return "No Security Reviewer rule violations detected in the fixture diff.";
  }

  return findings
    .map(
      (finding) =>
        `[${finding.severity}] ${finding.title}\n  rule: ${finding.rule}\n  path: ${finding.path}\n  ${finding.detail}`,
    )
    .join("\n\n");
}
